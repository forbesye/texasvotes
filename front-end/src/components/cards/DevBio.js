/*
    This file defines a card that displays each developer
    on the About page. Cards have a circular image, name, 
    and description.
*/

import React from "react"
import { Card, Typography } from "antd"
import styles from "./Card.module.css"
import {
	BranchesOutlined,
	ProfileOutlined,
	CheckCircleOutlined,
} from "@ant-design/icons"
const { Meta } = Card

// Icons from Ant Design for Git
const statIcons = {
	commits: BranchesOutlined,
	issues: ProfileOutlined,
	tests: CheckCircleOutlined,
}

// Git icon and number from Git
const GitStat = ({ type, n }) => {
	const Icon = statIcons[type]
	return (
		<div className={styles.gitStat}>
			<div>{type}</div>
			<Icon />
			<div>{n}</div>
		</div>
	)
}

/*
    Card to display for developer description

    Props:
        name (string) - developer's name
        bio (string) - developer's bio
        role (string) - front-end/back-ned
        picture_path (string) - path to developer's image
        commits (int) - number of Git commits
        issues (int) - number of Git issues
        tests (int) - number of Git tests
*/
const DevBio = (props) => {
    
    // Unpack the props
	const { name, bio, role, picture_path, commits, issues, tests } = props

	return (
		<Card
            bordered={true} 
            // Top of card; image of developer
			cover={
				<img
					className={styles.circle_headshot}
					alt={name}
					src={picture_path}
				/>
			}
            className={styles.card}
            // Match height of other components
			bodyStyle={{
				alignItems: "stretch",
				height: "100%",
				display: "flex",
				flexDirection: "column",
			}}
		>
            {/* Developer's name */}
			<Typography.Title level={3}>{name}</Typography.Title>
            {/* Developer's role and bio*/}
			<Meta title={role} description={bio} className={styles.meta} />
            {/* Git statistics at bottom of card */}
			<div className={styles.gitStats}>
				<GitStat type="commits" n={commits} />
				<GitStat type="issues" n={issues} />
				<GitStat type="tests" n={tests} />
			</div>
		</Card>
	)
}

export default DevBio
