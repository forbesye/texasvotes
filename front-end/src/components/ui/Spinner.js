import React from "react"
import { Spin } from "antd"

const messages = [
	"Looking for the fabled blue wave",
	"Waiting for Ken Paxton to stand trial for securities fraud",
	"Listening to Dennis Bonnen lose his job as Speaker of the House",
	"Dreaming of a Biden/Harris upset",
	"Waiting for Governor Greg Abbott to terrorize liberal cities once more",
	"Waiting for Lt. Governor Dan Patrick to say something stupid again",
	"Attempting to court the Hispanic vote",
	"Looking for Beto O'Rourke stumping on some tables",
	"Trying to stop Ted Cruz from reading 'Green Eggs and Ham' on the Senate Floor",
	"Searching for the Zodiac Killer",
	"Stalling the USPS",
	"Brushing off cheeto dust",
]

const styles = {
	margin: "auto",
	display: "block",
}

/**
 * Returns an Ant Design spinner component with a randomized loading message
 */
export default function Spinner() {
	const message =
		messages[Math.floor(Math.random() * messages.length)] + "..."
	return <Spin style={styles} tip={message} size="large" />
}
