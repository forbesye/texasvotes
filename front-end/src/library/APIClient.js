import axios from "axios"
import cache from "lru-cache"

/**
 * Configures an axios client to link to our back-end API using env. var.
 */
const client = axios.create({
	baseURL: process.env.REACT_APP_API_URL
	? process.env.REACT_APP_API_URL
	: "https://apidev.texasvotes.me"
})
const apiCache = new cache()
/**
 * API client from axios, allows modularity across models and is configured to
 * be linked to our back-end API
 * @param {model, path, params} elements
 */
const getAPI = async ({ model, path, params }) => {
	const url = path ? `/${model}/${path}` : `/${model}`
	const config = params ? { params: params } : {}
	const hash = `${url}${config?.params.toString()}`
	if(apiCache.has(hash)) {
		return apiCache.get(hash)
	}
	else {
		const result = await client.get(url, config)
		const { data } = result
		apiCache.set(hash, data)
		return data
	}
}

export { getAPI }
