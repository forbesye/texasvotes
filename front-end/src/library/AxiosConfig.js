import axios from "axios"

const api_url = process.env.REACT_APP_API_URL
	? process.env.REACT_APP_API_URL
	: "https://apidev.texasvotes.me"

const client = axios.create({
	baseURL: api_url,
})

export default client
