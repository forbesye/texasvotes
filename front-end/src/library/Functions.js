/*
 * Returns a string, only use for final output!
 */
const numberStringWithCommas = (num) => {
	return num.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",")
}

/**
 * Returns MM-DD-YYYY date string
 * @param {string} d date in ISO format
 */
const monthDayYearParse = (d) => {
	const date = new Date(d)
	return date.toLocaleString("en-US", {
		month: "short",
		day: "numeric",
		year: "numeric",
	})
}

/**
 * Returns a function to update the params with the appropriate filter
 * @param {f()} filter 
 * @param {f()} setParams 
 * @param {T} params 
 */
const updateFilter = (filter, setParams, params) => {
	const output = (value) => {
		setParams({
			...params,
			page: 1,
			[filter]: value
		})
	}
	return output
}

/**
 * Returns a percentage from val and total
 * @param {Number} val 
 * @param {Number} total 
 */
const convertToPercent = (val, total) => {
    return Math.round(
        (val / 100) * total
    )
}

/**
 * Returns the appropriate district phrase
 * @param {String} office 
 * @param {Number} number 
 */
const districtName = (office, number) => {
	const OFFICE_NAMES = {
		us_house: "US Congressional District",
		us_senate: "US Senate",
		tx_house: "Texas House of Representatives District",
		tx_senate: "Texas State Senate District",
	}

	if (office === "us_senate") {
		return "US Senate Seat for Texas"
	} else {
		return `${OFFICE_NAMES[office]} ${number}`
	}
}

/**
 * Returns number as string with format $XX.XX
 * @param {Number} num 
 */
const formatAsMoney = (num) => {
	return `$` + numberStringWithCommas(num.toFixed(2))
}


export {
	numberStringWithCommas,
	monthDayYearParse,
	convertToPercent,
	districtName,
	formatAsMoney,
	updateFilter
}
