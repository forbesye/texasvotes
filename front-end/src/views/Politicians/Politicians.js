import React, { Fragment, useState, useEffect } from "react"
import { Route, useLocation, useHistory } from "react-router-dom"
import { Typography, Divider, Tabs, Input } from "antd"
import SearchView from "./SearchView"
import GridView from "./GridView"
import styles from "./Politicians.module.css"

const { TabPane } = Tabs
const { Search } = Input

export default function Politicians () {
    const [currKey, setCurrKey] = useState("search")
    const history = useHistory()
    const location = useLocation()
    const handleChange = (key) => {
        history.push(`/politicians/${key}`)
        setCurrKey(key)
    }
    useEffect(() => {
        const path = location.pathname
        const initialKey = path.split("/").filter(el => el !== "").pop()
        setCurrKey(initialKey)
    }, [])

    return (
        <main className={styles.wrapper}>
            <header className={styles.header}>
                <Typography.Title level={1}>Texas Politicians</Typography.Title>
            </header>
            <div>
                <Tabs activeKey={currKey} onChange={handleChange}>
                    <TabPane tab="Search" key="search">
                        <Route path="/politicians/search">
                            <SearchView />
                        </Route>
                    </TabPane>
                    <TabPane tab="View All" key="view">
                        <Route path="/politicians/view">
                            <GridView />
                        </Route>
                    </TabPane>
                </Tabs>
            </div>
        </main>
    )
}