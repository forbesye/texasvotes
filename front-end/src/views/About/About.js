import React, { useEffect, useState } from 'react';
import DevBio from 'components/cards/DevBio'
import RepoCard from 'components/cards/RepoCard'
import ToolCard from 'components/cards/ToolCard'
import Spinner from 'components/ui/Spinner'
import styles from 'views/About/About.module.css'
import { toolsInfo, teamInfo, apiInfo, repoAndAPI } from "./AboutInfo"

const getGitlabInfo = async () => {
    let totalCommitCount = 0, totalIssueCount = 0, totalTestCount = 0;

    // Need to wipe member issues before calling again and calculate total tests
    teamInfo.forEach(member => {
        totalTestCount += member.tests;
        member.issues = 0;
    });

    // Can't use a map cause Gitlab's API returns are weird :/
    let commitList = await fetch("https://gitlab.com/api/v4/projects/21177395/repository/contributors")
    commitList = await commitList.json()
    commitList.forEach(element => {
        const { name, commits } = element;
        teamInfo.forEach(member => {
            if(member.name === name || member.username === name) {
                member.commits = commits;
            }
        })
        totalCommitCount += commits;
    });

    // Todo: When over 100 issues, implement pagination support to get all issues
    let issueList = await fetch("https://gitlab.com/api/v4/projects/21177395/issues?per_page=100")
    issueList = await issueList.json()
    
    issueList.forEach(element => {
        const { assignees } = element;
        // Todo: Check out what to do for multiple assignees
        assignees.forEach(a => {
            const { name } = a;
            teamInfo.forEach(member => {
                if(member.name === name || member.username === name) {
                    member.issues += 1;
                }
            })

            // if(teamInfo.has(name)) {
            //     teamInfo.get(name).issues += 1;
            // }
        });
        totalIssueCount += 1;
    })

    return {
        totalCommits: totalCommitCount,
        totalIssues: totalIssueCount,
        totalTests: totalTestCount,
        teamInfo: teamInfo
    }
}

const About = () => {
    const [teamList, setTeamList] = useState([]);
    const [totalCommits, setTotalCommits] = useState(0);
    const [totalIssues, setTotalIssues] = useState(0);
    const [totalTests, setTotalTests] = useState(0);
    const [ loaded, setLoaded ] = useState(false)

    useEffect(() => {
        const fetchData = async () => {
            if(teamList === undefined || teamList.length === 0) {
                const gitlabInfo = await getGitlabInfo();
                setTotalCommits(gitlabInfo.totalCommits);
                setTotalIssues(gitlabInfo.totalIssues);
                setTotalTests(gitlabInfo.totalTests);
                setTeamList(gitlabInfo.teamInfo);
                setLoaded(true);
            }
        }
        fetchData();
    }, [teamList])

    return(
        <div className={styles.wrapper}>
            <h1 className={styles.title} >About Us</h1>
                <p style={{fontSize: 20}}>
                TexasVotes is a website that allows users to quickly look up their representatives, the districts they live in, and state/federal elections they're slated to participate in within the state of Texas. We hope to increase governmental transparency and decrease the difficulty for the voting process in order to promote a more democratic society.
                </p>
            <h1 className={styles.title}>Our Team</h1>
            {
                loaded ?

                <div className={`${styles.gridLayout} ${styles.team}`}>
                {teamList.map(member => {
                    const { name, bio, role, picture_path, commits, issues, tests} = member;
                    return (
                        <DevBio 
                            key={name}
                            name={name}
                            role={role}
                            bio={bio}
                            picture_path={picture_path}
                            commits={commits}
                            issues={issues}
                            tests={tests}
                        />
                    )
                })}
                </div>

                : 
                <Spinner />
            }
            <h1 className={styles.title}>Repository Statistics</h1>
            <div className={`${styles.gridLayout} ${styles.repoStats}`}>
                <RepoCard type="commits" number={totalCommits}/>
                <RepoCard type="issues" number={totalIssues}/>
                <RepoCard type="tests" number={totalTests}/>
            </div>
            <h1 className={styles.title}>Development Tools</h1>
            <div className={`${styles.gridLayout} ${styles.devTools}`}>
                {toolsInfo.map(tool => {
                    const { title, img, description, link} = tool;

                    return (
                        <ToolCard
                            key={title} 
                            title={title}
                            img={img}
                            description={description}
                            link={link}
                        />
                    )
                })}
            </div>
            <h1 className={styles.title}>APIs Utilized</h1>
            <div className={`${styles.gridLayout} ${styles.devTools}`}>
                {apiInfo.map(api => {
                    const { title, img, description, link} = api;

                    return (
                        <ToolCard
                            key={title}
                            title={title}
                            img={img}
                            description={description}
                            link={link}
                        />
                    )
                })}
            </div>
            <h1 className={styles.title}>GitLab Repository and Postman API</h1>
            <div className={styles.flexbox}>
                {repoAndAPI.map(tool => {
                    const {img, link} = tool;

                    return (
                        <a href={link} key={link}>
                            <img className = {styles.logo} alt={link} src={img} />
                        </a>
                    )
                })}
            </div>
        </div>
    );

}

export default About