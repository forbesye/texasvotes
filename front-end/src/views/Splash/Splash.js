import React, { useState, useEffect, useRef } from "react"
import { Card, Typography, Button } from "antd"
import { DownOutlined } from "@ant-design/icons"
import { Link, useHistory } from "react-router-dom"

// Icons made by <a href="https://www.flaticon.com/authors/good-ware" title="Good Ware">Good Ware</a> from <a href="https://www.flaticon.com/" title="Flaticon"> www.flaticon.com</a>
import texas from "views/Splash/images/texas.svg"

// Icons made by <a href="https://www.flaticon.com/authors/freepik" title="Freepik">Freepik</a> from <a href="https://www.flaticon.com/" title="Flaticon"> www.flaticon.com</a>
import politician from "views/Splash/images/politician.svg"
import vote from "views/Splash/images/vote.svg"

import styles from "./Splash.module.css"
import GeneralSearchBar, { AddressSearchBar } from "../Search/GeneralSearchBar"
import News from "./News"
import { getAPI } from "library/APIClient"

const { Title, Paragraph } = Typography

const Splash = () => {
	const history = useHistory()
	const ref = useRef(null)
	const [ generalSearch, setGeneralSearch ] = useState(true)
	const [ searchVal, setSearchVal ] = useState("")
	const [ address, setAddress ] = useState("")
	const [ news, setNews ] = useState({
		articles: [],
		loading: false,
		lastUpdated: null,
	})

	useEffect(() => {
		const getNews = async () => {
			setNews((news) => ({ ...news, loading: true }))
			const { articles, last_updated } = await getAPI({ model: "news" })
			setNews({
				articles: articles,
				loading: false,
				lastUpdated: new Date(last_updated),
			})
		}
		getNews()
	}, [])

	return (
		<div className={styles.splashPage}>
			<div className={styles.splash}>
				<div className={styles.splashContent}>
					<div className={styles.headerText}>
						<Title level={1} className={styles.title}>
							Texas Votes
						</Title>
						<Paragraph className={styles.about}>
							Become informed about Texas politics today. Search
							for information about a politician, a district, or
							an election below.
						</Paragraph>
					</div>
					<div className={styles.searchSelectButtons}>
						<Button 
							ghost
							className={styles.searchSelectButton}
							onClick={() => setGeneralSearch(true)}
						>Search By Query</Button>
						<Button 
							ghost
							className={styles.searchSelectButton}
							onClick={() => setGeneralSearch(false)}
						>Search By Address</Button>
					</div>
					{ generalSearch ?
						<GeneralSearchBar
							clear
							onSearch={(value) => history.push(`/search?q=${value}`)}
							onChange={(text) => setSearchVal(text)}
							value={searchVal}
						/> :
						<AddressSearchBar 
							clear
							placeholder="Enter address here."
							onSearch={(value) => history.push(`/search/address?q=${value}`)}
							onChange={(text) => setAddress(text)}
							onOptionSelect={(val) => setAddress(val)}
							value={address}
						/>
					}
				</div>
				<div className={styles.downButton}>
					<DownOutlined
						className={styles.downButtonIcon}
						onClick={() => {
							window.scrollTo({
								top: ref.current.offsetTop,
								behavior: "smooth",
							})
						}}
					/>
				</div>
			</div>
			<div className={styles.modelCards} ref={ref}>
				<Title level={1} style={{ textAlign: "center" }}>
					Model Pages
				</Title>
				<div className={styles.cardFlexContainer}>
					<Link
						id="politicianCard"
						to="/politicians/view"
						className={styles.card}
					>
						<Card
							bordered={true}
							className={styles.card}
							cover={
								<img
									className={styles.cardImage}
									alt={"Politician"}
									src={politician}
								></img>
							}
							hoverable={true}
						>
							<Title level={2} style={{ textAlign: "center" }}>
								Politicians
							</Title>
							<Paragraph>
								Want to learn more about your politicians?
							</Paragraph>
						</Card>
					</Link>
					<Link
						id="districtCard"
						to="/districts/view"
						className={styles.card}
					>
						<Card
							bordered={true}
							className={styles.card}
							cover={
								<img
									className={styles.cardImage}
									alt={"Texas"}
									src={texas}
								></img>
							}
							hoverable={true}
						>
							<Title level={2} style={{ textAlign: "center" }}>
								Districts
							</Title>
							<Paragraph>
								Which district are you in? Learn about the
								different districts across Texas!
							</Paragraph>
						</Card>
					</Link>
					<Link
						id="electionCard"
						to="/elections/view"
						className={styles.card}
					>
						<Card
							bordered={true}
							className={styles.card}
							cover={
								<img
									className={styles.cardImage}
									alt={"Vote"}
									src={vote}
								></img>
							}
							hoverable={true}
						>
							<Title level={2} style={{ textAlign: "center" }}>
								Elections
							</Title>
							<Paragraph>
								What are some of the upcoming elections? Keep
								track of the elections and mark your calendars
								to vote!
							</Paragraph>
						</Card>
					</Link>
				</div>
			</div>
			<div className={styles.newsContainer}>
				<Title level={1}>Texas News Articles</Title>
				<Paragraph
					style={{
						textAlign: "center",
					}}
				>
					Last Updated:{" "}
					{news.lastUpdated
						? news.lastUpdated.toLocaleString("en-US")
						: "Never"}
				</Paragraph>
				<News {...news} />
			</div>
		</div>
	)
}

export default Splash
