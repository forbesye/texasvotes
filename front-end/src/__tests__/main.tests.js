import React from "react"
import { configure, shallow } from "enzyme"
import Adapter from "enzyme-adapter-react-16"
import About from "views/About/About"
import Splash from "views/Splash/Splash"
import DevBio from "components/cards/DevBio"
import Spinner from "components/ui/Spinner"
import Politicians from "views/Politicians/Politicians"
import { numberStringWithCommas, monthDayYearParse } from "library/Functions"

configure({ adapter: new Adapter() })

describe("Render views", () => {
	test("About", () => {
		const aboutTest = shallow(<About />)
		expect(aboutTest).toMatchSnapshot()
	})

	test("Splash", () => {
		const splashTest = shallow(<Splash />)
		expect(splashTest).toMatchSnapshot()
	})

	test("Politicians", () => {
		const politicianTest = shallow(<Politicians />)
		expect(politicianTest).toMatchSnapshot()
	})
})

describe("Render components", () => {
	test("Dev card", () => {
		const devCard = shallow(
			<DevBio
				name="ayy lmao"
				bio="ayy lmao"
				role="ayy lmao"
				picture_path="https://i.kym-cdn.com/photos/images/newsfeed/000/092/076/1293393293093.jpg"
				commits={2020}
				issues={2020}
				tests={2020}
			/>
		)
		expect(devCard).toMatchSnapshot()
	})

	test("Spinner", () => {
		const spinner = shallow(<Spinner />)
		expect(spinner).toMatchSnapshot()
	})
})

describe("Library functions", () => {
	test("Commas", () => {
		const num = 123456789
		const commas = jest.fn((num) => numberStringWithCommas(num))
		commas(num)
		expect(commas).toHaveReturnedWith("123,456,789")
	})
})
