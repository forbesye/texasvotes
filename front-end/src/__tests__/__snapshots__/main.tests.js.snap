// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render components Dev card 1`] = `ShallowWrapper {}`;

exports[`Render components Spinner 1`] = `ShallowWrapper {}`;

exports[`Render views About 1`] = `ShallowWrapper {}`;

exports[`Render views Splash 1`] = `ShallowWrapper {}`;
